### PR TITLE
Raaz windows support

### LINKED ISSUE #280: System random for windows
Support system prg for windows platform.

### PR DESCRIPTION
**Pull request by Mistuke**
This should fix #280 and #281

All tests pass.

### CODE REVIEW

File: entropy/windows/Raaz/Entropy.hsc [hunk 1]
```diff
@@ -30,18 +31,19 @@ type HCRYPTPROV = Ptr ()
...
+foreign import WINDOWS_CCONV unsafe "Wincrypt.h CryptAcquireContextW"
     c_CryptAcquireContext :: Ptr HCRYPTPROV -> CWString -> CWString
                           -> Word32 -> Word32 -> IO Bool
 
 foreign import WINDOWS_CCONV unsafe "Wincrypt.h CryptReleaseContext"
     c_CryptReleaseContext :: HCRYPTPROV -> Word32 -> IO Bool
 
 -- | Cache the crypto context so we don't have to create it on each call.
-cryptoContext :: ForeignPtr (HCRYPTPROV)
+cryptoContext :: ForeignPtr HCRYPTPROV
```
Comment (piyush-kurur) [inline entropy/windows/Raaz/Entropy.hsc]: I think you can avoid this code. I understand that you want to cache the crypto context for performance but note that the system entropy is only used to seed the internal chacha20 based CSPRG. So calls to getEntropy is done every 64GB or so.

A similar situation occurs in the `/dev/urandom` code where we open a handle to `/dev/urandom` for every call of getEntropy. You can follow that model.

Comment (piyush-kurur) [inline entropy/windows/Raaz/Entropy.hsc]: Also you may disable the -v in the appveyor builds so that we can get to the error quickly.
See appveyor.yml

Comment (Mistuke) [inline entropy/windows/Raaz/Entropy.hsc]: Ah, I wasn't aware it's only a seed. Fair enough, I'll update the patch.

Comment (piyush-kurur) [inline entropy/windows/Raaz/Entropy.hsc]: This library exposes the command raaz which can be used to generate random bytes. So you may try the windows equivalent of this 

```bash
stack exec raaz rand -- 10 | od -x
```
and inspect the output. Surprisingly this simple test caught an error on the OpenBSD system. There is also the script in the subdirectory  `./spec/prgquality.sh` which runs the dieharder on the output of stack exec raaz rand. If dieharder is available on windows you can try that (If you do not mind that is).

Comment (Mistuke) [inline entropy/windows/Raaz/Entropy.hsc]: Don't seem to have dieharder but generating random numbers seems to work correctly

```
Tamar@Rage /c/U/T/D/P/g/raaz> raaz rand -- 10 | od -x
0000000 4a90 c4b7 1ac0 6524 0478
0000012
Tamar@Rage /c/U/T/D/P/g/raaz> raaz rand -- 10 | od -x
0000000 14ce f1ba a478 195a 70ad
0000012
Tamar@Rage /c/U/T/D/P/g/raaz> raaz rand -- 10 | od -x
0000000 f613 7bf3 4549 a039 1e5e
0000012
Tamar@Rage /c/U/T/D/P/g/raaz> raaz rand -- 10 | od -x
0000000 d005 46ad 0b9a d148 0ecc
0000012
Tamar@Rage /c/U/T/D/P/g/raaz> raaz rand -- 10 | od -x
0000000 8a21 70cc 10c8 6c49 3159
0000012
```

File: entropy/windows/Raaz/Entropy.hsc [hunk 2]
```diff
@@ -38,28 +35,22 @@ foreign import WINDOWS_CCONV unsafe "Wincrypt.h CryptAcquireContextW"
...
+       ctx'    <- peek addr
        success <- c_CryptGenRandom ctx' (fromIntegral bytes) (castPtr ptr)
-       return $ if success then BYTES bytes else BYTES 0
-  where BYTES bytes = inBytes l
+       free addr
+       _ <- c_CryptReleaseContext ctx 0
+       if success
+          then return $ BYTES bytes
+          else fail "Unable to generate entropy. Call to CryptGenRandom failed."
+  where BYTES bytes = inBytes l
```
Comment (piyush-kurur) [inline entropy/windows/Raaz/Entropy.hsc]: Sorry to be pedantic about formating. Please add new line here.

File: entropy/windows/Raaz/Entropy.hsc [hunk 3]
```diff
@@ -38,28 +35,22 @@ foreign import WINDOWS_CCONV unsafe "Wincrypt.h CryptAcquireContextW"
...
-freeContext ctx = c_CryptReleaseContext ctx 0 >> return ()
-
 -- | Get cryptographically random bytes from the system.
 getEntropy :: (MonadIO m, LengthUnit l) => l -> Pointer -> m (BYTES Int)
-getEntropy l ptr = liftIO $ withForeignPtr cryptoContext $ \ctx ->
-    do ctx' <- peek ctx
+getEntropy l ptr = liftIO $
+    -- You can't get an address of a pointer in haskell, so making out
+    -- parameters are messy. We do this by boxing the type.
+    do ctx    <- mallocBytes (#size HCRYPTPROV)
```
Comment (piyush-kurur) [inline entropy/windows/Raaz/Entropy.hsc]: For the context you need a `Ptr  something` (may be `Ptr Word8`). Is that correct?  There is a combinator called [`allocaBuffer`](https://www.stackage.org/haddock/lts-8.9/raaz-0.1.1/Raaz-Core-Types.html) in the module `Raaz.Core.Types`. It is a safer (and often faster) form of malloc and free. There are variants that take care alignment restrictions as well. 

These family of allocation functions are built on top of the [`allocaBytes`](https://hackage.haskell.org/package/base-4.9.1.0/docs/Foreign-Marshal-Alloc.html)  function  from the `Foreign.Marshall.Alloc` modules.

**Comment by piyush-kurur** [comment]
The apveyor is giving an error that says undefined VirtualUnlock. Is there any extra options that needs to be passed to the ghc for linking?

File: entropy/windows/Raaz/Entropy.hsc (hunk 3 above)
Comment (piyush-kurur) [inline entropy/windows/Raaz/Entropy.hsc]: I guess the code outline should be like this.

```haskell
getEntropy = allocaBuffer (sizeOf (undefined :: Ptr HCRYPTPROV)) \ ptr -> 
     do _ <- peek ptr nullPtr
         cryptoAcquireCxt (castPtr ptr) ....
         cxt <- peek (castPtr ptr)
         do the reading
          cryptoReleaseCxt cxt
         



```

Just for reference here is what the entropy package does.

https://github.com/TomMD/entropy/blob/master/System/EntropyWindows.hs

**Comment by Mistuke** [comment]
No, it's part of Kernel32 which is already being passed. But it seems the 32 bit package of GCC doesn't contain `VirtualLock` In it's definitions. ugh. I'll write a stub for the 32 bit builds.

**Comment by piyush-kurur** [comment]
On Sun, Jun 25, 2017 at 04:20:43AM -0700, Tamar Christina wrote:
> No, it's part of Kernel32 which is already being passed. But it seems the 32 bit package of GCC doesn't contain `VirtualLock` In it's definitions. ugh. I'll write a stub for the 32 bit builds.

So apveyor only has 32-bit windows? is that true?

Regards,

ppk

**Comment by Mistuke** [comment]
I've never set up AppVeyor before but yeah every instance I've seen of it only has 32bit GHCs. which unfortunately the 32 bit GCC isn't of the same quality as the 64 bit build, as hardly anyone uses it. So we have to work around such issues. I usually patch them upstream but that of course won't fix the current issue.

**Comment by piyush-kurur** [comment]
On Sun, Jun 25, 2017 at 04:29:22AM -0700, Tamar Christina wrote:
> I've never set up AppVeyor before but yeah every instance I've seen of it only has 32bit GHCs. which unfortunately the 32 bit GCC isn't of the same quality as the 64 bit build, as hardly anyone uses it. So we have to work around such issues. I usually patch them upstream but that of course won't fix the current issue.

Can you hop on to #haskell on freenode irc. I am online we can discuss this online.
There might be others there who  can help.

**Comment by piyush-kurur** [comment]
Just to let you know that currently I have set both the apeveyor builds to be allow_failures. So if you see a green on appveyor do not get too happy ;-) make sure to check the builds. Once your pull request is successfully merged, then I will make the appveyor build count.

**Comment by Mistuke** [comment]
Alright. I haven't forgotten, just had to quickly switch to finishing
another patch. Will finish this this week!

**Comment by piyush-kurur** [comment]
No worries. Take your time. I just warned you about the appveyor stuff.

File: entropy/windows/Raaz/Entropy.hsc (hunk 3 above)
Comment (Mistuke) [inline entropy/windows/Raaz/Entropy.hsc]: Yeah, `entropy` has a bug though, as they're using `Word32` for `Ptr ()` which is obviously wrong on 64-bit platforms :)

**Comment by Mistuke** [comment]
That should do it, AppVeyor passes for both 32 and 64 bit now (though you should just disable -v). and Linux passes but MacOS seems stuck, though that seems an Travis issue rather than one with this package?

**Comment by piyush-kurur** [comment]
@mistuke that was awesome. I will merge it some time. including the cleanup of the appveyor.

**Pull request merged**